File: 01073. Adding Two Negabinary Numbers/solution.go
/**
 * Problem: 1073. Adding Two Negabinary Numbers
 *
 * Difficulty: Medium
 *
 * Language: Golang
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

func addNegabinary(arr1 []int, arr2 []int) []int {
		// If arr1 is shorter than arr2, swap them so arr1 is always the longer one
		if len(arr1) < len(arr2) {
			arr1, arr2 = arr2, arr1
		}
		
		// Add arr2 to arr1 from least significant bit to most significant bit
		for i := 0; i < len(arr2); i++ {
			arr1[len(arr1)-i-1] += arr2[len(arr2)-i-1]
		}

		// Initialize result slice to store the final answer
		result := []int{}

		// Normalize arr1 to handle negabinary carries
		for i := 0; i < len(arr1); i++ {
			index := len(arr1)-i-1 // Current index from right to left
			currentBit := arr1[index] // Current bit value
			arr1[index] %= 2 // Reduce bit to 0 or 1

			// If currentBit is 0 or 1, no carry needed
			if currentBit <= 1 {
				continue
			}
			// If previous bit exists and is positive, subtract carry from it
			if index-1 >= 0 && arr1[index-1] > 0 {
				arr1[index-1]--
			// Otherwise, propagate carry to previous two bits
			} else if index-2 >= 0 {
				arr1[index-1]++
				arr1[index-2]++
			// If no previous bits, prepend [1, 1] to result
			} else {
				result = []int{1,1}
			}
		}

		// Append normalized arr1 to result
		result = append(result, arr1...)

		// Find the first non-zero bit to remove leading zeros
		var firstOneIndex int
		for firstOneIndex = 0; firstOneIndex < len(result); firstOneIndex++ {
			if result[firstOneIndex] == 1 {
				break
			}
		}

		// If all bits are zero, return [0]
		if len(result) == firstOneIndex {
			return []int{0}
		}
		
		// Return result without leading zeros
		return result[firstOneIndex:]
}

File: 01073. Adding Two Negabinary Numbers/solution.js
/**
 * Problem: 1073. Adding Two Negabinary Numbers
 *
 * Difficulty: Medium
 *
 * Language: JavaScript
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

/**
 * Adds two negabinary numbers (base -2) as arrays (MSB first)
 *
 * @param {number[]} arr1 - First negabinary number (MSB first)
 * @param {number[]} arr2 - Second negabinary number (MSB first)
 *
 * @returns {number[]} - Sum in negabinary (MSB first)
 */
const addNegabinary = (arr1, arr2) => {
  // Initialize pointers for arr1 and arr2 to their last indices (LSB position)
  let i = arr1.length - 1,
    j = arr2.length - 1
  // Array to store the result digits (LSB first)
  const result = []

  // Variable to keep track of the carry in negabinary addition
  let carry = 0

  // Loop until all digits and carry are processed
  while (i >= 0 || j >= 0 || carry !== 0) {
    // Get current digit from arr1 or 0 if out of bounds
    const a = i >= 0 ? arr1[i] : 0,
      // Get current digit from arr2 or 0 if out of bounds
      b = j >= 0 ? arr2[j] : 0
    // Calculate sum of digits and carry
    const total = a + b + carry

    // Push the current digit (total mod 2) to result
    result.push(total & 1)

    // Update carry for next iteration (negabinary logic)
    carry = -(total >> 1)

    // Move to next digit (towards MSB)
    i--
    j--
  }

  // Remove leading zeros from result (except for single zero)
  while (result.length > 1 && result[result.length - 1] === 0) result.pop()

  // Reverse result to convert from LSB-first to MSB-first order
  return result.reverse()
}
